### DISCUSSION TITLE
Hardware instancing again, but for gltfio

### DISCUSSION THREAD

**Discussion by gibssa** [google/filament#6847] [Q&A] [answered]
Hi again Filament team,

`gltifo` is missing an API for [hardware instancing](https://github.com/google/filament/blob/321777e8c1e4a5a608295f7f4175c5552652dcc3/filament/include/filament/RenderableManager.h#L420-L433).

**Problem**

- Thanks for [your previous help](https://github.com/google/filament/discussions/6739) on manual instancing. 
  - I got a basic sphere instanced via vertex shader.
  - Metal profiler confirms reduction from 100 draw calls -> 1 draw call.
- `gltfio`'s [`FilamentInstance`](https://github.com/google/filament/blob/aa9f0a4ca3b09759da73788b9837bcf5d22890eb/libs/gltfio/include/gltfio/AssetLoader.h#L159-L184) != GPU instancing.
  - Metal profiler shows many repeated draw calls.

**Solution?**

- Is it reasonable to use [an approach similar to the internal `Wireframe.h`](https://github.com/google/filament/blob/321777e8c1e4a5a608295f7f4175c5552652dcc3/libs/gltfio/src/Wireframe.cpp#L45-L156)?
  - like this?
```
  mVertexBufer = // somehow copy from `gltfio` output
  mIndexBuffer = // somehow copy from `gltfio` output

  RenderableManager::Builder(1)
      .instances(100)
      .geometry(0, RenderableManager::PrimitiveType::TRIANGLES, mVertexBuffer, mIndexBuffer)
      .build(*engine, mEntity);
```

Or is there a better solution to get hardware instancing working on `gltfio` models?

**Comment by bejado** [comment, marked as answer]
I don't think gltfio will support hardware instancing without some work, but if you're building Filament from source you can hack this. Essentially what you want to do though is call `.instances` on `RenderableManager::Builder` each time gltfio creates a Renderable. We merged a new API a few weeks ago:

```
Builder& instances(size_t instanceCount, InstanceBuffer* instanceBuffer) noexcept;
```

This allows you to supply an `InstanceBuffer` with transforms to automatically offset the position of each instance, so you don't need to change the vertex shader.

**Reply by gibssa** [reply]
I am indeed building Filament from source. I'll try out an approach like this and close this out if I can figure it out.

Thanks again team!

**Reply by gibssa** [reply]
(deleted my previous response, was half-baked)

For `RenderableManger.Builder.instances(instance_buffer)`, it looks like `InstanceBuffer` supports dynamic transforms at runtime:

```
void setLocalTransforms(math::mat4f const* localTransforms, size_t count, size_t offset = 0);
```

But for my use case, I need to set other uniforms dynamically too (e.g. `baseColorFactor`).

Does the Filament team have any plans to either extend the `InstanceBuffer` struct to support more uniforms, or alternatively have an OKR or similar down the road for adding custom vertex shaders / similar hacks to get hardware instancing in gltfio?

(thanks again, as always!)

**Reply by bejado** [reply]
We don't have any plans for this at the moment. If you want to pass other per-instance uniforms, I recommend creating a parameter array in your material, for example:

```
{
    type : float4[16],
    name : baseColorFactor
}
```

Then you can index into it using `getInstanceIndex()`: `materialParams_baseColorFactor[getInstanceIndex()]`.

**Reply by gibssa** [reply]
Sounds good, no worries.

I am indeed passing in a parameter array for a couple of my non-glTF instanced Renderables (per advice in #6739). I'll pick up where I left off on your initial advice to modify `gltfio` to call `.instances()`. I'll try to take this a step further and also wire up `gltfio` to load some of my `filamat` materials that I throw into the build.

Thanks again!

**Reply by gibssa** [reply]
Got it working with your advice to directly hack `gltfio`, by passing in both `.instances()` and my own `.material()`.

Thanks @bejado 🙏
